util/flocks/src/tfrwlock/mod.rs

```diff
@@ -18,39 +20,176 @@ pub enum TfrwLockError {
     FileError(#[from] std::io::Error),
 }
 
-impl From<FrwLockError> for TfrwLockError {
-    fn from(e: FrwLockError) -> Self {
-        match e {
-            FrwLockError::LockNotAvailable => TfrwLockError::LockNotAvailable,
-            FrwLockError::FileError(e) => TfrwLockError::FileError(e),
-        }
+impl From<tokio::sync::TryLockError> for TfrwLockError {
+    fn from(e: tokio::sync::TryLockError) -> Self {
+        TfrwLockError::LockNotAvailable
     }
 }
 
-/// A file-based read-write lock.
-/// This only mutually excludes processes trying to violate the lock, not the same process--which is not considered contention.
-/// If you want to prevent contention within the same process, you should wrap this in your preferred synchronization primitive.
+/// Wraps a file-based read-write lock in a Tokio-friendly interface.
 pub struct TfrwLock<T: AsFd> {
-    lock : RwLock<FrwLock<T>>
+    lock : RwLock<T>
 }
 
 impl<T: AsFd> TfrwLock<T> {
+    
     pub fn new(file: T) -> Self {
         Self {
-            lock: RwLock::new(FrwLock::new(file))
+            lock: RwLock::new(file)
+        }
+    }
+
+    pub(crate) fn try_write(&self) -> Result<TfrwLockWriteGuard<'_, T>, TfrwLockError> {
+
+        let (res, write) = {
+            let file = self.lock.try_write()?;
+            (
+                flock(&*file, FlockOperation::NonBlockingLockExclusive),
+                file
+            )
+        };
+
+        match res {
+            Ok(_) => {
+                Ok(TfrwLockWriteGuard {
+                    guard : write
+                })
+            },
+            Err(rustix::io::Errno::WOULDBLOCK) => Err(TfrwLockError::LockNotAvailable),
+            Err(e) => Err(TfrwLockError::FileError(e.into())),
+        }
+
+    }
+
+    pub(crate) fn try_read(&self) -> Result<TfrwLockReadGuard<'_, T>, TfrwLockError> {
+        
+        let (res, read) = {
+            let file = self.lock.try_read()?;
+            (
+                flock(&*file, FlockOperation::NonBlockingLockShared),
+                file
+            )
+        };
+
+        match res {
+            Ok(_) => {
+                Ok(TfrwLockReadGuard {
+                    guard : read
+                })
+            },
+            Err(rustix::io::Errno::WOULDBLOCK) => Err(TfrwLockError::LockNotAvailable),
+            Err(e) => Err(TfrwLockError::FileError(e.into())),
+        }
+
+    }
+
+    pub async fn write(&self) -> Result<TfrwLockWriteGuard<'_, T>, TfrwLockError> {
+        loop {
+            match self.try_write() {
+                Ok(guard) => return Ok(guard),
+                Err(TfrwLockError::LockNotAvailable) => {
+                    // Yield control to other tasks
+                    yield_now().await;
+                },
```

Comment:
This will busy-loop when the file is locked by another process. To play nice with the OS, you'd need to use a blocking `flock`. To make it properly async-friendly, this lacks a way to wake up the task when locking status on the file changes. I'm not aware of any polling API that could be used for this purpose, though.

Reply:
We could always sleep a little and just assert this is a low-performance lock. Some number of microseconds could be reasonable as file writes should be on that order. 

I thought about using `inotify`, for which there are async stream implementations. But, I figured that might wind up being be heavier and it requires paths.

POSIX `aio` isn't well-supported ofc.

Reply:
> inotify may help if you use the file presence as the lock or "weak-lock" (i.e. the lock holder must unlock and remove the file). But this is less flexible and adds another failure mode: a process may die and leave the file, so inotify will never trigger.

Placing it on `xattrs` would be a bit better wouldn't it? Either way, yes, you would still need to poll as well. 

> This may be something that can only be solved by spawn_blocking.
You mean just for the thread pools? You don't think it's overkill. `flock` is blocking, but it should be a short operation, no? It just checks some bytes in an inode, no?

Reply:
If `flock` is called in the blocking mode, it's blocked while the lock is held. That's the dilemma: if it's blocking and the call waits on the lock in async context, the whole thread is blocked in the work-stealing runtime. If it's non-blocking, there is no nice way to schedule a wakeup for the next good time to poll for the lock, so you either busy-loop or sleep; each of these strategies have their own disadvantages.

`spawn_blocking` is not that expensive, but it might be pending longer and less predictably because the thread pool may be fully booked.

Reply:
I actually wasn't even considering not using `LOCK_NB`.

When the lock actually gets acquired, that's always a blocking operation, but a very brief one, right? I mean it has to be atomic at least. `LOCK_NB` is just to do non-blocking while checking the lock, right?

My inclination would be to just do `LOCK_NB` with polling and accept the small overhead for the block when the lock is actually acquired. But, I guess I don't really mind doing `spawn_blocking` if you think that's the right way to go. Just want a little more info here.

Reply:
> When the lock actually gets acquired, that's always a blocking operation, but a very brief one, right?

Yes, uncontested operation is not of concern.

> My inclination would be to just do `LOCK_NB` with polling

If you can find an efficient way to wake up the task, that would be ideal for async. `xattrs` with `inotify`? This is not portable, adds a persistent failure mode if the xattr is not cleaned up by the holding process, and imposes some extra requirements on the file system (to think of it, `flock` may do as well, but that's in POSIX and is dependent only on the OS kernel support).